src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,422 @@
+#include "postgres.h"
+#include "libpq-fe.h"
+#include "executor/spi.h"
+#include "distributed/lock_graph.h"
+#include "distributed/coordinator_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_logical_replication.h"
+#include "postmaster/postmaster.h"
+#include "distributed/distribution_column.h"
+#include "utils/builtins.h"
+#include "distributed/shard_split.h"
+#include "utils/lsyscache.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/background_jobs.h"
+#include "distributed/multi_join_order.h"
+#include "distributed/citus_ruleutils.h"
+
+PG_FUNCTION_INFO_V1(citus_auto_shard_split_start);
+PG_FUNCTION_INFO_V1(citus_find_shard_split_points);
+
+int64 MaxShardSize = 102400;
+double TenantFrequency = 0.3;
+
+/*
+ * Struct to store all the information related to
+ * a shard.
+ */
+typedef struct ShardInfoData
+{
+	int64 shardSize;
+	int64 shardId;
+	int32 nodeId;
+	int64 shardGroupSize;
+}ShardInfoData;
+typedef ShardInfoData *ShardInfo;
+
+void ErrorOnConcurrentOperation(void);
+StringInfo GetShardSplitQuery(ShardInfo shardinfo, Datum splitPointArrayDatum,
+							  char *shardSplitMode);
+void ExecuteSplitBackgroundJob(int64 jobId, ShardInfo shardinfo, Datum splitPointArrayDatum,
+							   char *shardSplitMode);
+List * FindShardSplitPoints(int64 shardId);
+bool ScheduleShardSplit(ShardInfo shardinfo, char *shardSplitMode, int64 jobId);
+
+/*
+ * It throws an error if a concurrent automatic shard split or Rebalance operation is happening.
+ */
+void
+ErrorOnConcurrentOperation()
+{
+	int64 jobId = 0;
```

Comment:
we may need a lock here to handle concurrent calls to citus_auto_shard_split (see AcquireRebalanceColocationLock), since HasNonTerminalJobOfType once the job is already running